Huge Sprite Rework

ktccd opened the pull request:
Reverts Citadel-Station-13/Citadel-Station-13#340
:cl: ktccd
fix: Fixed how Taur sprites are layered, finally making genitals show for Nagas as they should.
tweak: Vaginas and Breasts now follow the same naming convention for sprites that penises use, so all genital organs can be rendered using the same piece of code. This means that breast sizes have dropped to single-letter variants, since those are what we got sprites for (Because the cup-size is now the Size of the obj).
tweak: Tweaked how colours are chosen (Humans can enable/disable matching genitals to
their skintones, everyone else has to pick their colours).
tweak: Breasts now have shapes! So far we only got the "pair" shape, but we got code-support for other variants.
add: Vaginas can now also be chosen and will render properly, meaning we now have more of them to choose from (Some of them terrifying). They don't really use a size, but there's code support for it if sprites would be made for it.
tweak: Breast items now change their size and colour to the owner's when/if they somehow lose the organs (gibbing for example). Both Breasts and Vaginas change their description to match what the object is set to look like, sprites are lacking in that area however.
tweak: Minor variable error in the penis-description code fixed.
fix: Taur sprites renamed to use the correct layers. Shephard sprite bug fixed where there was a missing 2 pixels in the middle of the body.
/:cl:

[why] # 
Our custom spriting was a bit of a mess and should be working now. There's still some stuff that can be done and improved upon, such as adding testicles onmob sprites and more variants, shapes, sizes etc. But now we got code support for it, so adding more shapes should be fairly simple, and adding more sizes even simpler.
I'll attach a changelog yml file to this in the next commit, since I don't know the number of my PR before making one.

ktccd committed:
Revert "Revert "Huge Sprite Rework""

TalkingCactus commented on code/citadel/organs/genitals.dm:
```diff
@@ -198,21 +208,28 @@
-		if(H.has_vagina())
-			genitals_to_add += H.getorganslot("vagina")
+		else
+			if(H.has_vagina()) //Sadly, a penis would cover the vagina and they use the same layer, so don't draw both.
```
You can draw more than one overlay per layer, it simply stacks them on top of one another in the order they were applied. Remove or clarify this comment, please.

ktccd replied:
Yeah, I'm not sure what the order this would result in.
Do we WANT both to be drawn one over the other? I was a bit paranoid about drawings things being some sort of costly action after talking to TG coderbus about images...

TalkingCactus replied:
Overlays applied first will be on the bottom. So do vagina -> testicles(when we do get sprite support) -> penis

ktccd replied:
✔️ Vaginas are now drawn under penises on the same layer.

TalkingCactus commented on code/citadel/organs/genitals.dm:
```diff
@@ -198,21 +208,28 @@
 			switch(G.type)
 				if(/obj/item/organ/genital/penis)
 					S = cock_shapes_list[G.shape]
 					size = G.size
```
`size = G.size` can be moved above the switch statement and removed from each individual type since it's generalized for all genitals.

ktccd replied:
True, this was kind of a hacky tacked-on just to make it work, before I made more things include a size in the imagename ^^.

ktccd replied:
✔️

TalkingCactus commented on code/citadel/organs/genitals_sprite_accessories.dm:
```diff
@@ -27,15 +27,57 @@
+
+//I think this one is intended for arousal... but it would only work for human vaginas in this case...
+//So I'm leaving it as a selectable option until we can make/need variants for the sprites to change based on arousal.
+/datum/sprite_accessory/vagina/dripping
```
Your comment was correct. This effect was intended for an overlay and/or through examine text on the human mob; remove it or provide support for it as an overlay.

ktccd replied:
Is there a difference between a sprite_accessory and an overlay?
My hope would be to have one basic and one aroused version for each type and use a genital boolean to check which one to use.
Although I think I could use place-holder copies for the ones missing images right now.

TalkingCactus replied:
An overlay is an image/icon that is 'overlayed' onto the base icon of the /atom that it's applied to. A `/datum/sprite_accessory` is a holder of information, like color source and icon file/state, for a specific overlay.

ktccd replied:
alright, thanks ^^. Tying this to the questions I had on the other similar review change.

ktccd replied:
✅ Reworked arousal for genitals to support **possible** alternate icon_state overlays for arousal. Defaults to not using them, meaning they look the same both aroused and not.

TalkingCactus commented on code/citadel/organs/genitals_sprite_accessories.dm:
```diff
@@ -27,15 +27,57 @@
+	icon_state = "human"
+	name = "Human"
+
+/datum/sprite_accessory/vagina/gaping
```
Same situation as `.../dripping`.

ktccd replied:
I think I'll just remove this for now then, since this seems like it'd need it's own framework, while the wetness ones could be sort-of tacked on to the arousal framework.

ktccd replied:
✔️ Locked  = 1 due to missing supporting framework

Iamgoofball replied:
:/

TalkingCactus commented on code/citadel/organs/vagina.dm:
```diff
@@ -5,13 +5,60 @@
+	icon_state = sanitize_text(string)
+	var/lowershape = lowertext(shape)
+
+	desc = "That's a [lowershape] vagina. You estimate it could stretch about [round(cap_length, 0.25)] inch[cap_length > 1 ? "es" : ""] deep, around something [round(cap_girth, 0.25)] inch[cap_girth > 1 ? "es" : ""] thick \
```
`You estimate it to be around [round(cap_length, 0.5)] inch[cap_length > 1 ? "es" : ""] deep and could stretch around something [round(cap_girth, 0.5)] inch[cap_girth > 1 ? "es" : ""] in girth. It has [clits > 1 ? "[clits] clits" : "a clit"] about [round(clit_len,0.25)] inch[clit_len > 1 ? "es" : ""] long and [round(clit_diam, 0.25)] inch[clit_diam > 1 ? "es" : ""] in diameter.`

Fixes some grammatical errors. May be imperfect as I'm only reading it from here instead of in game.

ktccd replied:
✔️

TalkingCactus commented on code/citadel/organs/vagina.dm:
```diff
@@ -5,13 +5,60 @@
+	else
+		desc += "not very wet."
+	if(owner)
+		color = "#[owner.dna.features["vag_color"]]"
```
Color is assigned upon creation during character spawning. We don't want to force an arbitrary color onto an organ simply because their preferences call for it. Removing this will prevent issues later on when genitals can be grown to specs for surgical implanting. If a genital is regrown or otherwise regenerated, then we would apply their DNA's color preference.

ktccd replied:
but isn't this in the update_appearance proc? Meaning we run this proc when we **want** to update the appearance?
I admit I'm a bit unsure when the proc is supposed to be used, I might've even stolen it from the penis code completely.

TalkingCactus replied:
This proc exists to set the name/icon_state/desc/etc upon initialization or whenever else we need to.

ktccd replied:
The color thing was copied from the penis.dm code of the same proc, so I assumed this proc was meant to also update colors. Still not sure why this is wrong?

ktccd replied:
Ok, so I checked this a bit.
If we don't have these lines, we end up with those white-uncolored sprites for genitals when you gib someone. because there is no assigning color code upon creation during character spawning.
EDIT: Or if there is, it doesn't take skin_tone into consideration.
EDIT2: Now that I re-checked a bit, I think this is why I had all those code-pieces in the preferences that reset genital colors when you changed skin-color. because otherwise the genital colors stay as whatever.
I think putting it in the update code like this might be the easiest and most sensible way to update it to work? Like, it sounds like this is exactly what the proc is meant to do anyway....

ktccd replied:
✅ Disputed change, though the testing lead me to improve the code in a different way.
The update appearance proc, as written even before I fixed it, is meant to be used to force the genitals to update the appearance. this CAN cause radical changes, like changing a human shape into a knotted shape, so it should only be called when such a transformation is wanted.

TalkingCactus commented on code/citadel/organs/vagina.dm:
```diff
@@ -5,13 +5,60 @@
+		if("dripping")
+			desc += "\nIt is gaping slightly, inflamed and "
+		else
+			desc += "\nIt has an exotic shape and is "
```
This switch() should be run before line 27 and inserted through `[ ]`s instead of tacked on afterwards.

ktccd replied:
Yeah, I admit IDK fully how the [] things really work, which is why I used a lot of desc += things to build a description.

ktccd replied:
✔️

TalkingCactus replied:
`[ ]` allows you to embed variables or procs into text. Look through the docs with F1 in dream maker for things like this.

ktccd replied:
Yeah, I was able to implement it through testing my way through until I got how it works. As it is now, I use the switch-case to build up a temp variable that is later put in the desc using [ ] at the end.

TalkingCactus commented on code/modules/client/preferences.dm:
```diff
@@ -405,6 +406,8 @@ var/list/preferences_datums = list()
 					dat += "<b>Tertiary Color: </b><span style='border: 1px solid #161616; background-color: #[features["mcolor3"]];'>&nbsp;&nbsp;&nbsp;</span> <a href='?_src_=prefs;preference=mutant_color3;task=input'>Change</a><BR>"
 				if(pref_species.use_skintones)
 					dat += "<b>Skin Tone: </b><a href='?_src_=prefs;preference=s_tone;task=input'>[skin_tone]</a><BR>"
+					dat += "<b>Genitals use skintone:</b><a href='?_src_=prefs;preference=genital_colour'>[features["genitals_use_skintone"] == TRUE ? "Enabled" : "Disabled"]</a><BR>"
```
`Genitals use skintone` -> `Genitals Use Skintone`

ktccd replied:
In the description text?

TalkingCactus replied:
Yes, it needs capitals to match the rest of the character creation screen since it's the title of a preference.

ktccd replied:
✔️

TalkingCactus commented on code/modules/client/preferences.dm:
```diff
@@ -1338,6 +1369,22 @@ var/list/preferences_datums = list()
+							features["cock_color"] = sanitize_hexcolor(skintone2hex(skin_tone))
+							features["breasts_color"] = sanitize_hexcolor(skintone2hex(skin_tone))
+							features["vag_color"] = sanitize_hexcolor(skintone2hex(skin_tone))
+							features["balls_color"] = sanitize_hexcolor(skintone2hex(skin_tone))
```
The skin tone preference is applied in the `give_[genital]()` series of procs. You can remove these color changes. Same with the `else` clause below and on lines `1182 - 1185` and `861 - 865`.

ktccd replied:
These are in place so that the preferences page updates the selected colours shown. If you toggle the skintone option, it will now change the colour on the right to match the skintone.
In hindsight, the s_tone at 1182 needs an if-statement so it only updates if the option is enabled, like I did for line 861.

TalkingCactus replied:
The shown color preview can be changed to their skin tone or just hidden if the option is enabled.

TalkingCactus replied:
With this, their custom colors are lost if they toggle the preference, too.

ktccd replied:
so you mean the actual colour in their features is the same as before, but it should **display** a skintone to the user instead? I think I can do that. Should end up making the skintone able to toggle off and on that way also, since right now it only toggles TO skintone and kind of sticks there when you toggle off and need manual change....

ktccd replied:
✔️

TalkingCactus commented on code/modules/client/preferences_savefile.dm:
```diff
@@ -148,10 +148,12 @@ SAVEFILE UPDATING/VERSIONING - 'Simplified', or rather, more coder-friendly ~Car
 		features["breasts_fluid"] 		= sanitize_inlist(features["breasts_fluid"], milk_id_list, "milk")
 		//vagina features
 		features["has_vag"]				= sanitize_integer(features["has_vag"], 0, 1, 0)
+		features["vag_shape"]			= sanitize_inlist(features["vag_shape"], vagina_shapes_list, "human")
```
On line 5, change the max version to `19` and change the `if()` on line 128 to `if(current_version < 19)` or these will not sanitize.

ktccd replied:
✔️

TalkingCactus commented on code/citadel/organs/genitals_sprite_accessories.dm:
```diff
@@ -27,15 +27,57 @@
+
+/datum/sprite_accessory/vagina/hairy
+	icon_state = "hairy"
+	name = "Hairy"
```
This should be a flag, variable, or a separate preference, like hair and beards, so it can be applied as an overlay

ktccd replied:
My worry with that is that it might be incompatible with the other shapes? Logically, you'd think you could have it on top of the shape, but some of them don't look like they lend themselves to it easily.
I imagine this wouldn't look very good overlayed on top of, say dentata or tentacles, since it's drawn at the same location as the other ones.
In fact, the fact that it's drawn exactly where the others are, seemed to me that the sprite isn't "hairy" as much as it looks "furry"? ((I admit it's hard to tell what the intention is, with the size of the sprites, but I know at least the coords of the sprite pixels means it will be right on top of the other ones, which I thought would interfere with some))

ktccd replied:
✅ Disputed, mostly because it doesn't fit with most of the sprites. If we want customisable hair-styles as some sort of additional overlay, we'll need some sprites for that too. As it is, I'm only adding this in as a shape, because that's what the icon we got is drawn as (it would completely cover most of the other sprites, making them mutually exclusive anyway).

TalkingCactus commented on code/citadel/organs/breasts.dm:
```diff
@@ -31,3 +31,23 @@
+	else
+		desc += "\nThey do not seem to be producing liquids."
+	if(owner)
+		color = "#[owner.dna.features["breasts_color"]]"
```
Refer to my comment about color changing in `vagina.dm`.

ktccd replied:
✅ same

TalkingCactus reviewed (changes requested):
This is what I found on my first pass, sorry it took so long.

ktccd committed:
Compilation of all feedback fixes

This commit contains all the suggested fixes so far and the changes we
came up with through discussion about the code.
For example, genitals **may** now use alternate icons for aroused
states, but does not have to, this is decided by a boolean variable in
the sprite accessory itself, defaulting to not using alternate icons.
Genital icon_state names now follow a new format to match these new
options in a modular way for all kinds of genitals.

TalkingCactus commented on code/citadel/organs/vagina.dm:
```diff
@@ -23,30 +22,37 @@
-			desc += "\nIt has an exotic shape and is "
-	if(wetness)
-		desc += "slick with female arousal."
+			details = "\nIt has an exotic shape and is "
```
Why are you putting these on new lines?

ktccd replied:
Uhm.... because I thought a description would look really cluttered?
IDK, it was a pretty vague notion, does it look weird?

ktccd committed:
Conflict solving

reformats all the taurs to use the TAUR layer instead of the BODY_FRONT
layer. Same for the new Eevee sprites.

ktccd committed:
Committing the old, unfixed file.

This will cause errors if used, but at least it won't cause conflicts.

ktccd committed:
Contains fixed sprite names

The shepherd taur body should STILL have a hole in the body due to a
pixel error.
I have forgotten which pixes these are, so.... **le shrug**

ktccd committed:
Anger commit

Commiting this in anger. I WILL make it work damnit!

ktccd committed:
Proper name fixes.

Also renames the taur cow to no longer have an incorrectly duplicate
icon_state.

ktccd committed:
Shepherd taur pixel fix

Fixes the missing pixels from the taur body that caused a hole to appear
when viewed from the sides.

ktccd committed:
Conflict resolution

This PR will cause some sprite-breaking overlay errors.
But it's the only way to handle the conflicting .dmi.
So I'll make a separate PR to fix the sprites later.

ktccd committed:
Merge branch 'master' into revert-340-revert-339-Genital_Overlays

ktccd committed:
Fixes global lists

Now they use whatever weird new system we got for them.
